File: 446. Arithmetic Slices II - Subsequence/Solution.cpp
typedef long long ll;
class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& nums) {
        vector<unordered_map<ll,ll>>m(nums.size(),unordered_map<ll,ll>());
        ll t=0;
        for (int i=1;i<nums.size();i++){
            for(int j=0;j<i;j++){
                ll d=1ll*nums[i]-1ll*nums[j];
                t+=m[j][d];
                m[i][d]=m[i][d]+m[j][d]+1;
            }
        }
        return (int)t;
    }
};

File: 446. Arithmetic Slices II - Subsequence/Solution.py
class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        t=0
        m=[dict() for i in range(len(nums))]
        for i in range(1,len(nums)):
            for j in range(i):
                d=nums[i]-nums[j]
                t+=m[j][d] if d in m[j] else 0
                m[i][d]=(0 if d not in m[i] else m[i][d])+(0 if d not in m[j] else m[j][d])+1
        return t
